src/java.base/share/classes/java/lang/reflect/ProxyGenerator.java

```diff
@@ -440,71 +414,80 @@ private static void collectCompatibleTypes(Class<?>[] from,
         }
     }
 
-    /**
-     * Returns the {@link ClassLoader} to be used by the default implementation of {@link
-     * #getCommonSuperClass(String, String)}, that of this {@link ClassWriter}'s runtime type by
-     * default.
-     *
-     * @return ClassLoader
-     */
-    protected ClassLoader getClassLoader() {
-        return loader;
-    }
-
     /**
      * Generate a class file for the proxy class.  This method drives the
      * class file generation process.
      */
     private byte[] generateClassFile() {
-        visit(CLASSFILE_VERSION, accessFlags, dotToSlash(className), null,
-                JLR_PROXY, typeNames(interfaces));
+        var localCache = new HashMap<ClassDesc, ClassHierarchyResolver.ClassHierarchyInfo>();
+        return Classfile.build(classDesc, List.of(Classfile.Option.classHierarchyResolver(classDesc ->
```

Comment:
How/when is `ClassHierarchyResolver` being called?   I didn't dig deeper.

Reply:
`ClassHierarchyResolver` is used for stack maps generation in situations when it is not clear if the given type is class or interface and to resolve class hierarchy (parent - child relations and common parent).
Classfile API is not using class loaders by default (In contrast to ASM library) and there are alternative ways how to provide required info. However in this situation is class hierarchy resolution asking provided class loader.